sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExpressionCodegen.scala

```diff
@@ -0,0 +1,259 @@
+      case BoundReference(ordinal, _, _) if ctx.currentVars == null ||
+          ctx.currentVars(ordinal) == null =>
+        Seq(ctx.INPUT_ROW)
+
+      // An expression which is not evaluated yet. Tracks down to find input rows.
+      case BoundReference(ordinal, _, _) if ctx.currentVars(ordinal).code != "" =>
+        trackDownRow(ctx, ctx.currentVars(ordinal))
+
+      case _ => Seq.empty
+    }.distinct
+  }
+
+  /**
+   * Tracks down input rows referred by the generated code snippet.
+   */
+  def trackDownRow(ctx: CodegenContext, exprCode: ExprCode): Seq[String] = {
+    var exprCodes: List[ExprCode] = List(exprCode)
+    val inputRows = mutable.ArrayBuffer.empty[String]
+
+    while (exprCodes.nonEmpty) {
+      exprCodes match {
+        case first :: others =>
+          exprCodes = others
+          if (first.inputRow != null) {
+            inputRows += first.inputRow
+          }
+          first.inputVars.foreach { inputVar =>
+            if (inputVar.exprCode.code != "") {
+              exprCodes = inputVar.exprCode :: exprCodes
+            }
+          }
+        case _ =>
+      }
+    }
+    inputRows.toSeq
+  }
+
+  /**
+   * Retrieves previously evaluated columns referred by children and deferred expressions.
+   * Returned tuple contains the list of expressions and the list of generated codes.
+   */
+  def getInputVarsForChildren(
+      ctx: CodegenContext,
+      expr: Expression): (Seq[Expression], Seq[ExprCode]) = {
+    expr.children.flatMap(getInputVars(ctx, _)).distinct.unzip
+  }
+
+  /**
+   * Given a child expression, retrieves previously evaluated columns referred by it or
+   * deferred expressions which are needed to evaluate it.
+   */
+  def getInputVars(ctx: CodegenContext, child: Expression): Seq[(Expression, ExprCode)] = {
+    if (ctx.currentVars == null) {
+      return Seq.empty
+    }
+
+    child.flatMap {
+      // An evaluated variable.
+      case b @ BoundReference(ordinal, _, _) if ctx.currentVars(ordinal) != null &&
+          ctx.currentVars(ordinal).code == "" =>
+        Seq((b, ctx.currentVars(ordinal)))
+
+      // An input variable which is not evaluated yet. Tracks down to find any evaluated variables
+      // in the expression path.
+      // E.g., if this expression is "d = c + 1" and "c" is not evaluated. We need to track to
+      // "c = a + b" and see if "a" and "b" are evaluated. If they are, we need to return them so
+      // to include them into parameters, if not, we track down further.
+      case b @ BoundReference(ordinal, _, _) if ctx.currentVars(ordinal) != null =>
+        trackDownVar(ctx, ctx.currentVars(ordinal))
+
+      case _ => Seq.empty
+    }.distinct
+  }
+
+  /**
+   * Tracks down previously evaluated columns referred by the generated code snippet.
+   */
+  def trackDownVar(ctx: CodegenContext, exprCode: ExprCode): Seq[(Expression, ExprCode)] = {
+    var exprCodes: List[ExprCode] = List(exprCode)
```

Comment:
what about making this a mutable collection to avoid its complete copy here https://github.com/apache/spark/pull/19813/files#diff-0fcf78140c2c4edef3840822ac41563eR204?